### PR TITLE
Support for specifying SSL certificates in Cassandra client

### LINKED ISSUE #3383: mTLS Support for Cassandra Cluster Connectivity
**Is your feature request related to a problem? Please describe.**
I'm looking to connect to a Cassandra cluster with mTLS. This requires an SSL certificate to be presented from the client-side.

**Describe the solution you'd like**
I would like to be able to specify paths to x509 encoded certificate and key files. The Cassandra client should use these files as part of its SSL configuration when communicating with the cluster.

**Describe alternatives you've considered**
A sidecar for adding the SSL encryption, but this seems a bit ridiculous given the driver supports this configuration.

**Additional context**
Sample logic showing how it is implemented - https://github.com/flightc/astra_gocql_connect/blob/master/astra_gocql_connect.go

### PR DESCRIPTION
**Pull request by bradfordcp**
**What this PR does**: Add support for Cassandra client SSL certificates

**Which issue(s) this PR fixes**:
Fixes #3383 

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by bradfordcp** [comment]
Completed the checklist and squashed the commits. Waiting for the checks to pass before submitting for review.

File: pkg/chunk/cassandra/storage_client.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ type Config struct {
 	SSL                      bool                `yaml:"SSL"`
 	HostVerification         bool                `yaml:"host_verification"`
 	CAPath                   string              `yaml:"CA_path"`
+	CertPath                 string              `yaml:"cert_path"`
+	KeyPath                  string              `yaml:"key_path"`
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:38]: We're trying to uniform config. Could you rename the YAML config options and their CLI flags respectively to:

- `tls_cert_path` / `tls-cert-path`
- `tls_key_path` / `tls-key-path`

File: pkg/chunk/cassandra/storage_client.go [hunk 2]
```diff
@@ -62,6 +64,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SSL, "cassandra.ssl", false, "Use SSL when connecting to cassandra instances.")
 	f.BoolVar(&cfg.HostVerification, "cassandra.host-verification", true, "Require SSL certificate validation.")
 	f.StringVar(&cfg.CAPath, "cassandra.ca-path", "", "Path to certificate file to verify the peer.")
+	f.StringVar(&cfg.CertPath, "cassandra.cert-path", "", "Path to certifcate file to present to cassandra cluster.")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:67]: `certifcate` > `certificate`

File: pkg/chunk/cassandra/storage_client.go [hunk 3]
```diff
@@ -86,6 +90,12 @@ func (cfg *Config) Validate() error {
 	if cfg.SSL && cfg.HostVerification && len(strings.Split(cfg.Addresses, ",")) != 1 {
 		return errors.Errorf("Host verification is only possible for a single host.")
 	}
+	if cfg.SSL && cfg.CertPath != "" && cfg.KeyPath == "" {
+		return errors.Errorf("Certificate specfied without key.")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:94]:
```suggestion
		return errors.Errorf("TLS certificate specified, but private key configuration is missing.")
```

File: pkg/chunk/cassandra/storage_client.go [hunk 4]
```diff
@@ -86,6 +90,12 @@ func (cfg *Config) Validate() error {
...
+	}
+	if cfg.SSL && cfg.KeyPath != "" && cfg.CertPath == "" {
+		return errors.Errorf("Key specified without certificate.")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:97]:
```suggestion
		return errors.Errorf("TLS private key specified, but certificate configuration is missing.")
```

File: pkg/chunk/cassandra/storage_client.go [hunk 5]
```diff
@@ -144,17 +154,29 @@ func (cfg *Config) setClusterConfig(cluster *gocql.ClusterConfig) error {
 	cluster.DisableInitialHostLookup = cfg.DisableInitialHostLookup
 
 	if cfg.SSL {
+		tlsConfig := &tls.Config{}
+
+		if len(cfg.CertPath) > 0 {
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:159]:
```suggestion
		if cfg.CertPath != "" {
```

File: pkg/chunk/cassandra/storage_client.go [hunk 6]
```diff
@@ -144,17 +154,29 @@ func (cfg *Config) setClusterConfig(cluster *gocql.ClusterConfig) error {
...
+			cert, err := tls.LoadX509KeyPair(cfg.CertPath, cfg.KeyPath)
+			if err != nil {
+				return errors.Wrap(err, "unable to load certificate and key")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:162]:
```suggestion
				return errors.Wrap(err, "unable to load TLS certificate and private key")
```

File: pkg/chunk/cassandra/storage_client_test.go [hunk 7]
```diff
@@ -75,6 +75,37 @@ func TestConfig_setClusterConfig_authWithPasswordAndPasswordFile(t *testing.T) {
...
+	assert.NotNil(t, cqlCfg.SslOpts)
+	assert.Len(t, cqlCfg.SslOpts.Certificates, 1)
+}
+
+func TestConfig_setClusterConfig_clientSSLWithOnlyCertificatePath(t *testing.T) {
+	cfg := defaultConfig()
+	cfg.SSL = true
+	cfg.CAPath = "testdata/example.com.ca.pem"
+	cfg.CertPath = "testdata/example.com.pem"
+	assert.Error(t, cfg.Validate())
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client_test.go:98]: Could you assert on the exact error message, please?

_Same comment applies to the next test._

**Review by pracucci — COMMENTED** [review]
Good job, thanks! I left few minor comments and then we should be good to go.

**Comment by bradfordcp** [comment]
I'll spend some cycles on getting this turned around tonight. Thanks for the review!

File: pkg/chunk/cassandra/storage_client.go [hunk 8]
```diff
@@ -86,6 +90,12 @@ func (cfg *Config) Validate() error {
 	if cfg.SSL && cfg.HostVerification && len(strings.Split(cfg.Addresses, ",")) != 1 {
 		return errors.Errorf("Host verification is only possible for a single host.")
 	}
+	if cfg.SSL && cfg.CertPath != "" && cfg.KeyPath == "" {
+		return errors.Errorf("TLS certificate specfied, but private key configuration is missing.")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:94]:
```suggestion
		return errors.Errorf("TLS certificate specified, but private key configuration is missing.")
```

File: pkg/chunk/cassandra/storage_client.go [hunk 9]
```diff
@@ -62,6 +64,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SSL, "cassandra.ssl", false, "Use SSL when connecting to cassandra instances.")
 	f.BoolVar(&cfg.HostVerification, "cassandra.host-verification", true, "Require SSL certificate validation.")
 	f.StringVar(&cfg.CAPath, "cassandra.ca-path", "", "Path to certificate file to verify the peer.")
+	f.StringVar(&cfg.CertPath, "cassandra.tls-cert-path", "", "Path to certificate file to present to cassandra cluster.")
+	f.StringVar(&cfg.KeyPath, "cassandra.tls-key-path", "", "Path to certificate key file for file as cassandra.cert-path.")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:68]:
```suggestion
	f.StringVar(&cfg.KeyPath, "cassandra.tls-key-path", "", "Path to private key file used by TLS.")
```

File: pkg/chunk/cassandra/storage_client.go [hunk 10]
```diff
@@ -62,6 +64,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SSL, "cassandra.ssl", false, "Use SSL when connecting to cassandra instances.")
 	f.BoolVar(&cfg.HostVerification, "cassandra.host-verification", true, "Require SSL certificate validation.")
 	f.StringVar(&cfg.CAPath, "cassandra.ca-path", "", "Path to certificate file to verify the peer.")
+	f.StringVar(&cfg.CertPath, "cassandra.tls-cert-path", "", "Path to certificate file to present to cassandra cluster.")
```
Comment (pracucci) [inline pkg/chunk/cassandra/storage_client.go:67]:
```suggestion
	f.StringVar(&cfg.CertPath, "cassandra.tls-cert-path", "", "Path to certificate file used by TLS.")
```

**Review by pracucci — APPROVED** [review]
LGTM, thanks! I left a couple of nits and we're good to go. Thanks for your patience!

**Comment by bradfordcp** [comment]
Thanks for your patience, I'm surprised by the number of typos :sweat_smile:

**Comment by pracucci** [comment]
> Thanks for your patience, I'm surprised by the number of typos 😅

Thanks to you @bradfordcp. LGTM. Let me ping another maintainer for the 2nd review and then we can merge it.

**Review by jtlisi — APPROVED** [review]
LGTM

Thank you for the PR!

**Pull request merged**